java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -221,6 +221,18 @@ abstract class AbstractArcHost(
      */
     protected suspend fun isArcHostIdle() = runningMutex.withLock { runningArcs.isEmpty() }
 
+    override suspend fun waitForArcIdle(arcId: String) {
```

Comment:
might be good to make this 
```
= withTimeout(SOME_CONSTANT_MS) {
  while (true) { 
    ...
   }
}
```

Just to prevent a resource leak (perma-running coroutine)

Reply:
if I were to do that, it would need to throw an exception if timing out (else waiting for idle wouldn't have any real meaning); I don't think that's the right approach though as it introduces brittleness (not to mention a single policy for all possible uses).

As demonstrated in testing, you can withTimeout your waitForArcIdle call in application code if you want to; I think that's probably sufficient.

Reply:
withTimeout() does throw an exception and cancels the coroutine (withTImeoutOrNull behaves differently), so the callers won't proceed as if the the idle was reached, they'd fail all the way up the call stack. 

 My worry is a scenario where someone's triggering ephemeral Arcs constantly in the background and waiting for idle (say, 1 per second), if there's some flake where some % of them get stuck, it'll end up as a silent resource leak. 

Maybe it'll be good to hear from @jasonwyatt. I feel like at least logging when extraordinarily long wait timeouts happening is a good idea, e.g. you could put the withTimeout() inside the while(true), and log() every time there's a timeout threshold exceeded (e.g. 30seconds, 10 seconds, whatever the threshold should be).  Same semantics as you have now, but some debugging info is admitted if things take overly long.  Android StrictMode has lots of logging like this. 

Could also parameterize the waitForArcIdle() with a maximum time (or 0 = infinite).

I approved it so you can merge, but it does kind of raise alarms and I just want to make sure everyone is aware, in case we do run into an issue later.

Reply:
we should definitely have the conversation.

I'm not opposed to e.g. Chronicle making the policy decision of setting a particular timeout threshold; and as I pointed out one of the tests demonstrates that this is feasible. My point is that the framework shouldn't be the thing that dictates what timeout policy is applied.